### PR TITLE
[feature] Dialogue Choice

### LINKED ISSUE #217: Implement Dialog Options

### PR DESCRIPTION
**Pull request by uciii**
### Main changes
This PR provides:
- Dialogue choice system
- Refactor dialogue state and action map into UI state and action map (for further UI development such as journal and etc could use this UI state and action map)
- Using event system for UI input action
- Delete all canvas in UI related prefab + wrap all UI prefab in some scene with canvas named UI + refactor their test code 

Note: use NPC intractable test scene to test this feature

### Tasks List
- [close #217] 

### Attachments

https://user-images.githubusercontent.com/71260420/157151607-4ce722f4-9587-4953-b2c2-aa9cdf56f503.mov

### CODE REVIEW

File: Assets/Astralization/Scripts/Player/PlayerStates/InputManager.cs [hunk 1]
```diff
@@ -71,13 +81,13 @@ public void HandleInputDefault(InputAction.CallbackContext ctx)
...
 
-    public void HandleInputDialogue(InputAction.CallbackContext ctx)
+    public void HandleInputUi(InputAction.CallbackContext ctx)
     {
         if (!CanHandleInput()) return;
-        switch (ctx.action.name)
-        {
-            case "NextDialogue": _currentPlayerState.NextDialogue(ctx); break;
-        }
+        //switch (ctx.action.name)
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/Player/PlayerStates/InputManager.cs:87]: remove this, handler function should be called from ui input button callbacks, for example pressing button

File: Assets/Astralization/Scripts/Player/PlayerStates/UiState.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+using System;
+using UnityEngine.InputSystem;
+
+public class UiState : PlayerState
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/Player/PlayerStates/UiState.cs:4]: class docs

**Review by rafiftaris — CHANGES_REQUESTED** [review]
Good implementation, just minor changes needed

**Review by rafiftaris — APPROVED** [review]
lgtm

**Review by naufalhilmi12345 — APPROVED** [review]
Cool! LGTM

File: Assets/Astralization/Scripts/Player/PlayerStates/InputManager.cs [hunk 3]
```diff
@@ -31,18 +31,28 @@ private void OnDisable()
     }
     #endregion
 
+    #region SetGet
     public void SetPlayerActionMap(Utils.PlayerHelper.States actionMap)
     {
-        _playerInput.SwitchCurrentActionMap(actionMap.ToString());
+        // for all UI related, add this condition
+        if (actionMap.ToString().Equals("Dialogue"))
```
Comment (raihansyahdean) [inline Assets/Astralization/Scripts/Player/PlayerStates/InputManager.cs:38]: I dont think this is necessary, we can instead change the states in the utils to "UI"

File: Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs [hunk 4]
```diff
@@ -4,7 +4,9 @@
 using UnityEngine.TestTools;
 using UnityEngine.InputSystem;
 using UnityEngine.SceneManagement;
-
+using UnityEngine.UI;
```
Comment (raihansyahdean) [inline Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs:7]: rename class and scene to VictimTest

File: Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs [hunk 5]
```diff
@@ -37,7 +39,7 @@ public IEnumerator PlayerInteractableDetector_InteractNPC()
         GameObject npc = GameObject.Find("NPC");
```
Comment (raihansyahdean) [inline Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs:39]: rename local variables to victim as well

File: Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs [hunk 6]
```diff
@@ -37,7 +39,7 @@ public IEnumerator PlayerInteractableDetector_InteractNPC()
...
         INpcController npcController = npc.GetComponent<INpcController>();
```
Comment (raihansyahdean) [inline Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs:40]: rename class to IVictimController and VictimController. Dont forget to adjust the scenes as well

**Review by raihansyahdean — CHANGES_REQUESTED** [review]

File: Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs (hunk 6 above)
Comment (raihansyahdean) [reply Assets/Astralization/Tests/NpcTest/NpcInteractableTest.cs:40]: the action inside the class should also be renamed: VictimInteractionEvent

**Review by raihansyahdean — APPROVED** [review]
Excellent

**Review by rafiftaris — APPROVED** [review]
kereeenn

**Pull request merged**